Automattic/jetmon
# Fix/site scanning on one minute rounds

This PR has a number of fixes for issues exposed after moving to one-minute rounds. Primarily it fixes the following:

1. The `localRetries` and `queuedRetries` arrays are no longer cleared on each round. This fixes the issue where site statuses could never be switched to down because the queues would reset before they could finish fully processing a site.
2. Fixed bucket range calculation issues in the `getNextBatch` function so that it no longer skips the final 64 buckets the host is configured to scan.
3. `gCountError`, `gCountOffline`, `gCountSuccess`, and `sitesCount` stats variables are no longer reset by `resetVariables`. Now they are reset after sending the stats so that the values don't consistently inflate over the course of a round and so that values aren't lost when the variables are reset before the stats are recorded. The values are cached in local vars to keep consistency in the stats logs and statsd.

There are some miscellaneous other changes included in this PR. Please let me know if I should move them to different PRs:

1. Added a `PEER_OFFLINE_LIMIT` config so that the value is no longer hard-coded in `lib/jetmon.js`. This value defaults to `3` as currently set in production if the value is missing in `config/config.json`.
2. Added a `DB_UPDATES_ENABLE` config and code toggled by this config that updates the `jetpack_monitor_sites` table with site status changes. This is very helpful for running local tests, but it should not be used in production. This defaults to disabled even if the config option is missing from `config/config.json`.
3. Added `config/config.readme` with short descriptions about the `config/config.json` settings.

## Testing Instruction

1. Ensure that your test environment's DB has a variety of testing sites that are up, down, very slow (2m+ to respond), and switching between up and down if possible. You should also have sites that have 5-minute and 1-minute `check_interval` in the DB.
2. Checkout this PR's branch: `fix/site-scanning-on-one-minute-rounds`.
3. Set `PEER_OFFLINE_LIMIT` to `1` and `DB_UPDATES_ENABLE` to `true` in `config/config.json`.
4. Run Jetmon.
5. Confirm that the database updates to properly reflect the status of each site. Note: depending on your configuration, it may take a few
6. Confirm that the sites are accessed at the check intervals that match their DB entry.
